File: PHP/src/OneLang/One/Transforms/InferTypesPlugins/ResolveFuncCalls.php
<?php

namespace OneLang\One\Transforms\InferTypesPlugins\ResolveFuncCalls;

use OneLang\One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\UnresolvedCallExpression;
use OneLang\One\Ast\Expressions\GlobalFunctionCallExpression;
use OneLang\One\Ast\Expressions\LambdaCallExpression;
use OneLang\One\Ast\References\GlobalFunctionReference;
use OneLang\One\Ast\AstTypes\LambdaType;

class ResolveFuncCalls extends InferTypesPlugin {
    function __construct() {
        parent::__construct("ResolveFuncCalls");
        
    }
    
    function canTransform($expr) {
        return $expr instanceof UnresolvedCallExpression;
    }
    
    function transform($expr) {
        $callExpr = $expr;
        if ($callExpr->func instanceof GlobalFunctionReference) {
            $newExpr = new GlobalFunctionCallExpression($callExpr->func->decl, $callExpr->args);
            $callExpr->args = array_map(function ($arg) { return $this->main->runPluginsOn($arg); }, $callExpr->args);
            $newExpr->setActualType($callExpr->func->decl->returns);
            return $newExpr;
        }
        else {
            $this->main->processExpression($expr);
            if ($callExpr->func->actualType instanceof LambdaType) {
                $newExpr = new LambdaCallExpression($callExpr->func, $callExpr->args);
                $newExpr->setActualType($callExpr->func->actualType->returnType);
                return $newExpr;
            }
            else
                return $expr;
        }
    }
}

File: Python/OneLang/One/Transforms/InferTypesPlugins/ResolveFuncCalls.py
from onelang_core import *
import OneLang.One.Transforms.InferTypesPlugins.Helpers.InferTypesPlugin as inferTypesPlug
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.References as refs
import OneLang.One.Ast.AstTypes as astTypes

class ResolveFuncCalls(inferTypesPlug.InferTypesPlugin):
    def __init__(self):
        super().__init__("ResolveFuncCalls")
    
    def can_transform(self, expr):
        return isinstance(expr, exprs.UnresolvedCallExpression)
    
    def transform(self, expr):
        call_expr = expr
        if isinstance(call_expr.func, refs.GlobalFunctionReference):
            new_expr = exprs.GlobalFunctionCallExpression(call_expr.func.decl, call_expr.args)
            call_expr.args = list(map(lambda arg: self.main.run_plugins_on(arg), call_expr.args))
            new_expr.set_actual_type(call_expr.func.decl.returns)
            return new_expr
        else:
            self.main.process_expression(expr)
            if isinstance(call_expr.func.actual_type, astTypes.LambdaType):
                new_expr = exprs.LambdaCallExpression(call_expr.func, call_expr.args)
                new_expr.set_actual_type(call_expr.func.actual_type.return_type)
                return new_expr
            else:
                return expr